File: src/lib/providers/provider-executor.mjs
// Sealed generic provider executor. Supports runner kinds: legacy-check,
// runtime-script, security-pack, external-process, imported-artifact, and
// reasoning-contract (planned/bundled but never invoked here). Internal module
// paths resolve relative to the installed Legion package and must match the
// sealed provider record.

import { pathToFileURL } from 'node:url';
import { resolve } from 'node:path';
import { runExternal } from './executor/external-process.mjs';
import { requireProjectExecutionSandbox } from '../host/sandbox-policy.mjs';
import { normalizeProviderResult } from './sdk/result.mjs';
import { validateProviderOutputAuthority } from './sdk/contracts.mjs';
import { validateSchema } from '../qualification/schema-validator.mjs';
import { readFile } from 'node:fs/promises';
import { createHash } from 'node:crypto';

export async function executePlannedProvider({
  provider,
  root,
  plan,
  projection,
  artifacts,
  options = {},
  host,
}) {
  requireProjectExecutionSandbox(provider, host);

  switch (provider.runner?.kind) {
    case 'legacy-check':
      return runLegacyCheck(provider, root, plan, projection, artifacts, host);
    case 'external-process':
      return runExternal({
        provider: provider.id,
        executable: provider.runner.executable,
        executableDigest: provider.runner.executableDigest,
        args: provider.runner.args ?? [],
        cwd: root,
        timeoutMs: plan.config?.limits?.providerTimeoutMs ?? 120000,
        maxOutputBytes: plan.config?.limits?.maxOutputBytes ?? 8388608,
        environmentKeys: provider.runner.environmentKeys ?? ['PATH', 'HOME'],
        projectExecution: provider.hostCapabilities?.includes('project-execution') === true,
      }, host);
    case 'runtime-script':
    case 'security-pack':
      return runRuntimeScript(provider, root, plan, projection, artifacts, host);
    case 'imported-artifact':
      return runImportedArtifact(provider, artifacts, host);
    case 'reasoning-contract':
      return runReasoning(provider,plan,projection,artifacts,host);
    default:
      return {
        provider: provider.id,
        status: 'error',
        complete: false,
        coverageGaps: [{ kind: 'unsupported-runner', runner: provider.runner?.kind ?? null }],
      };
  }
}

async function runLegacyCheck(provider, root, plan, projection, artifacts, host) {
  if(typeof host.processRunner?.run!=='function')return{provider:provider.id,status:'unproven',complete:false,coverageGaps:[{kind:'named-check-host-unavailable',check:provider.runner.check}]};
  const result=await host.processRunner.run({kind:'legion-named-check',provider:provider.id,check:provider.runner.check,cwd:root,binding:plan.binding??null,denominator:plan.denominator??null,projection,artifacts});
  if(!result||typeof result.complete!=='boolean')return{provider:provider.id,status:'unproven',complete:false,coverageGaps:[{kind:'named-check-receipt-invalid',check:provider.runner.check}]};
  return{provider:provider.id,status:result.status??'unproven',complete:result.complete===true,findings:result.findings??[],candidates:result.candidates??[],coverageGaps:result.coverageGaps??[]};
}

async function runReasoning(provider,plan,projection,artifacts,host){const review=host.reviewer?.review??host.reviewer?.run;if(typeof review!=='function')return{provider:provider.id,status:'unproven',complete:false,coverageGaps:[{kind:'reasoning-reviewer-unavailable'}]};const packet={schemaVersion:1,kind:'legion-reasoning-packet',provider:provider.id,contract:provider.runner.contract,binding:plan.binding??null,projection,artifactIds:[...artifacts.keys()]};const result=await review.call(host.reviewer,packet);if(!result||typeof result.complete!=='boolean')return{provider:provider.id,status:'unproven',complete:false,coverageGaps:[{kind:'reasoning-receipt-invalid'}]};return{provider:provider.id,status:result.status??'unproven',complete:result.complete===true,findings:result.findings??[],candidates:result.candidates??[],coverageGaps:result.coverageGaps??[]};}

async function runRuntimeScript(provider, root, plan, projection, artifacts, host) {
  const modulePath = resolveRepositoryModule(provider.runner.module??provider.runner.script);
  const moduleBytes=await readFile(modulePath);
  const moduleDigest=`sha256:${createHash('sha256').update(moduleBytes).digest('hex')}`;
  if(moduleDigest!==provider.runner.moduleDigest)throw new Error(`provider ${provider.id} module digest mismatch`);
  const module = await import(pathToFileURL(modulePath).href);
  if (typeof module.default === 'object' && module.default?.id && module.default.id !== provider.id) {
    throw new Error(`security pack ID mismatch: plan=${provider.id}, module=${module.default.id}`);
  }
  const runner = module.analyze ?? module.default?.analyze;
  if (typeof runner !== 'function') {
    throw new Error(`runtime-script ${provider.runner.module} exports no runnable function`);
  }
  const result = await runner({
    root,
    plan,
    projection,
    artifacts,
    provider: provider.id,
    host,
  });
  const normalized=normalizeProviderResult(provider,result);
  validateProviderOutputAuthority(provider,normalized);
  const schema=JSON.parse(await readFile(resolve(import.meta.dirname,'..','..','src/schemas/provider-result-v1.schema.json'),'utf8'));
  const issues=validateSchema(schema,normalized);
  if(issues.length)throw new TypeError(`provider ${provider.id} result schema invalid: ${issues.join(',')}`);
  return normalized;
}

async function runImportedArtifact(provider, artifacts, host) {
  const artifact = artifacts?.get(provider.runner.artifact);
  if (!artifact) {
    return {
      provider: provider.id,
      status: 'missing',
      complete: false,
      coverageGaps: [{ kind: 'imported-artifact-missing', artifact: provider.runner.artifact }],
    };
  }
  return {
    provider: provider.id,
    status: 'pass',
    complete: true,
    artifact: provider.runner.artifact,
  };
}

export function resolveRepositoryModule(script) {
  // Resolve scripts relative to the Legion installation root, never from a
  // mutable external registry path.
  if(!script)throw new Error('runtime-script requires sealed module path');
  if (script.startsWith('/') || /^[a-zA-Z]:[\\/]/.test(script)) {
    throw new Error(`runtime-script path must be repository-relative: ${script}`);
  }
  return resolve(import.meta.dirname, '..', '..', script);
}

File: engine/crates/legion-audit/src/wf_port/wf010/provider_executor.rs
//! Port of `src/lib/providers/provider-executor.mjs`'s dispatcher and its
//! `resolveRepositoryModule` path-safety guard.
//!
//! `legacy-check` and `reasoning-contract` are ported as a generic dispatch
//! over host-supplied async hooks (there is no native `host.processRunner`
//! / `host.reviewer` equivalent in this crate to call directly).
//!
//! `runtime-script` / `security-pack` (`runRuntimeScript`) and
//! `external-process` (`runExternal`, ported in
//! [`super::external_process::run_external`]) dispatch the same way the JS
//! does; `imported-artifact` is a pure function ported directly.
//!
//! GAP: JS `runRuntimeScript` performs a dynamic `import()` of a
//! JS-language provider module at a sealed, digest-pinned path, then invokes
//! its exported `analyze` function. There is no Rust analogue of loading and
//! invoking an arbitrary JS module at runtime; a native `runtime-script`
//! provider is instead a compiled Rust function the plan's runner record
//! names by key. That lookup/dispatch is out of this chunk's scope (it
//! lives with whatever registers native providers), so this port only
//! carries over `resolveRepositoryModule`'s path-safety check, which is
//! reusable by such a lookup.

use std::path::{Path, PathBuf};

use serde_json::Value;

/// Faithful port of `resolveRepositoryModule(script)`.
///
/// Rejects a `script` that is a POSIX absolute path (`/...`) or a Windows
/// drive-absolute path (`C:\...`), matching the JS regex
/// `/^[a-zA-Z]:[\\/]/`. `repository_root` stands in for the three
/// `resolve(import.meta.dirname, '..', '..', '..', script)` hops in JS,
/// i.e. the Legion package/repository root.
pub fn resolve_repository_module(script: &str, repository_root: &Path) -> Result<PathBuf, RepositoryModuleError> {
    if script.is_empty() {
        return Err(RepositoryModuleError::MissingModule);
    }
    let is_windows_drive_absolute = script.len() >= 3
        && script.as_bytes()[0].is_ascii_alphabetic()
        && script.as_bytes()[1] == b':'
        && matches!(script.as_bytes()[2], b'\\' | b'/');
    if script.starts_with('/') || is_windows_drive_absolute {
        return Err(RepositoryModuleError::NotRepositoryRelative { script: script.to_string() });
    }
    Ok(repository_root.join(script))
}

#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq)]
pub enum RepositoryModuleError {
    #[error("runtime-script requires sealed module path")]
    MissingModule,
    #[error("runtime-script path must be repository-relative: {script}")]
    NotRepositoryRelative { script: String },
}

/// Mirrors the minimal provider-result shape every branch of
/// `executePlannedProvider` returns on its unproven / missing / error paths.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct ProviderOutcome {
    pub provider: String,
    pub status: String,
    pub complete: bool,
    #[serde(default)]
    pub coverage_gaps: Vec<Value>,
    #[serde(default)]
    pub findings: Vec<Value>,
    #[serde(default)]
    pub candidates: Vec<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub artifact: Option<String>,
}

fn gap(kind: &str) -> Value {
    serde_json::json!({ "kind": kind })
}

fn gap_with(kind: &str, key: &str, value: &str) -> Value {
    serde_json::json!({ "kind": kind, key: value })
}

/// Faithful port of `runImportedArtifact(provider, artifacts, host)`.
///
/// `artifacts` mirrors the JS `Map<string, unknown>` keyed by artifact name;
/// only presence is inspected, matching JS (`artifacts?.get(...)` truthy
/// check).
pub fn run_imported_artifact(provider_id: &str, artifact_name: &str, artifacts: &std::collections::BTreeMap<String, Value>) -> ProviderOutcome {
    if !artifacts.contains_key(artifact_name) {
        return ProviderOutcome {
            provider: provider_id.to_string(),
            status: "missing".to_string(),
            complete: false,
            coverage_gaps: vec![gap_with("imported-artifact-missing", "artifact", artifact_name)],
            findings: vec![],
            candidates: vec![],
            artifact: None,
        };
    }
    ProviderOutcome {
        provider: provider_id.to_string(),
        status: "pass".to_string(),
        complete: true,
        coverage_gaps: vec![],
        findings: vec![],
        candidates: vec![],
        artifact: Some(artifact_name.to_string()),
    }
}

/// Faithful port of the `default:` branch of `executePlannedProvider`'s
/// `switch (provider.runner?.kind)`.
pub fn unsupported_runner(provider_id: &str, runner_kind: Option<&str>) -> ProviderOutcome {
    ProviderOutcome {
        provider: provider_id.to_string(),
        status: "error".to_string(),
        complete: false,
        coverage_gaps: vec![serde_json::json!({
            "kind": "unsupported-runner",
            "runner": runner_kind,
        })],
        findings: vec![],
        candidates: vec![],
        artifact: None,
    }
}

/// Faithful port of `runReasoning`'s / `runLegacyCheck`'s shared
/// "host hook unavailable" short-circuit: both bail out to an `unproven`
/// result naming the missing capability when the host cannot service the
/// request, rather than erroring.
pub fn host_hook_unavailable(provider_id: &str, gap_kind: &str) -> ProviderOutcome {
    ProviderOutcome {
        provider: provider_id.to_string(),
        status: "unproven".to_string(),
        complete: false,
        coverage_gaps: vec![gap(gap_kind)],
        findings: vec![],
        candidates: vec![],
        artifact: None,
    }
}

/// Faithful port of `runLegacyCheck`'s / `runReasoning`'s receipt-invalid
/// short-circuit (`typeof result.complete !== 'boolean'`), and their
/// success path otherwise, given the receipt already fetched by the caller
/// from its host hook (`host.processRunner.run` / `host.reviewer.review`).
pub fn from_host_receipt(provider_id: &str, receipt: Option<&Value>, invalid_gap_kind: &str) -> ProviderOutcome {
    let complete = receipt.and_then(|r| r.get("complete")).and_then(Value::as_bool);
    let Some(complete) = complete else {
        return ProviderOutcome {
            provider: provider_id.to_string(),
            status: "unproven".to_string(),
            complete: false,
            coverage_gaps: vec![gap(invalid_gap_kind)],
            findings: vec![],
            candidates: vec![],
            artifact: None,
        };
    };
    let receipt = receipt.unwrap();
    let status = receipt.get("status").and_then(Value::as_str).unwrap_or("unproven").to_string();
    let findings = receipt.get("findings").and_then(Value::as_array).cloned().unwrap_or_default();
    let candidates = receipt.get("candidates").and_then(Value::as_array).cloned().unwrap_or_default();
    let coverage_gaps = receipt.get("coverageGaps").and_then(Value::as_array).cloned().unwrap_or_default();
    ProviderOutcome {
        provider: provider_id.to_string(),
        status,
        complete,
        coverage_gaps,
        findings,
        candidates,
        artifact: None,
    }
}
